### PR TITLE
37607 rest configurable products SKU childrenendpoint missing media gallery entries

### LINKED ISSUE #37607: REST /configurable-products/sku/children endpoint missing media_gallery_entries
### Preconditions and environment

- Magento version: 2.4.3

### Steps to reproduce

1. Create a new configurable product and add configurations with unique product images. 
2. Make a request in postman using the `/configurable-products/{sku}/children`. 

### Expected result

Per the [rest api docs](https://developer.adobe.com/commerce/webapi/rest/quick-reference/?query=media_gallery_entries&keywords=&index=commerce-marketplace%2Ccommerce-contributor%2Ccommerce-cloud-tools%2Ccommerce-webapi%2Ccommerce-php%2Ccommerce-services%2Ccommerce-frontend-core%2Ccommerce-admin-developer%2Ccommerce-xd-kits%2Ccommerce-pwa-studio%2Cfranklin-commerce%2Ccommerce-testing%2Ccommerce-events) there should be a `media_gallery_entries` field on the response data for each child.

### Actual result

The `media_gallery_entries` field is missing entirely. Making a request to the `/products` endpoint searching for one of the configurations you will get a response containing the `media_gallery_entries` field but for some reason it is missing from the `/children` response. 

### Additional information

_No response_

### Release note

_No response_

### Triage and priority

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [X] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### PR DESCRIPTION
**Pull request by JettyRohanAditya**
media_gallery_entries field is missing from configurable-products/sku/children RestEndpoint

### Description (*)

media_gallery_entries field is missing in rest endpoint, Linkmanagement.php is the file for the rest endpoint.
That function is missing code for getting media_gallery_entries.
Created a custom factory for storing and displaying the media_gallery_entries. 

### Related Pull Requests

https://github.com/magento/magento2/issues/37607

### Fixed Issues (if relevant)

1. Fixes magento/magento2#37607

### Steps to reproduce
1. Create a new configurable product and add configurations with unique product images.
2. Make a request in postman using the /configurable-products/{sku}/children.
### Changes made:-
**Created custom Factory for media_gallery_entries.**

\Magento\Catalog\Api\Data\ProductAttributeMediaGalleryEntryInterfaceFactory $myModelFactory

**Created a custom function for setting media gallery entries into the factory and retrieved data from it.**

public function getMediaEntries($images){    $media = $this->myModelFactory->create();    $mediaGalleryEntries=[];    foreach ($images as $image) {        $media->setId($image["value_id"]);        $media->setMediaType($image["media_type"]);        $media->setLabel($image["label"]);        $media->setPosition($image["position"]);       $media->setDisabled($image["disabled"]);       $media->setFile($image["file"]);        $mediaGalleryEntries[]=$media->getData();    }    return $mediaGalleryEntries;} 

 **Line where we are getting all the media gallery entries from** 

$images= (array)$child->getMediaGallery('images'); 

**Line responsible for retrieving media gallery entries from the custom factory and giving them in result payload**

$attributes['media_gallery_entries'] = $this->getMediaEntries($images);

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by JettyRohanAditya** [comment]
@magento run all tests

**Comment by JettyRohanAditya** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

File: app/code/Magento/ConfigurableProduct/Model/LinkManagement.php [hunk 1]
```diff
@@ -47,6 +52,11 @@ class LinkManagement implements \Magento\ConfigurableProduct\Api\LinkManagementI
      */
     private $attributeFactory;
 
+    //private $mediaGalleryProcessor;
```
Comment (engcom-Hotel) [inline app/code/Magento/ConfigurableProduct/Model/LinkManagement.php]: I think we can remove it from here as it is of no use

File: app/code/Magento/ConfigurableProduct/Model/LinkManagement.php [hunk 2]
```diff
@@ -55,24 +65,47 @@ class LinkManagement implements \Magento\ConfigurableProduct\Api\LinkManagementI
...
      */
     public function __construct(
         \Magento\Catalog\Api\ProductRepositoryInterface $productRepository,
         \Magento\Catalog\Api\Data\ProductInterfaceFactory $productFactory,
         \Magento\ConfigurableProduct\Model\ResourceModel\Product\Type\Configurable $configurableType,
         \Magento\Framework\Api\DataObjectHelper $dataObjectHelper,
-        \Magento\Catalog\Model\ResourceModel\Eav\AttributeFactory $attributeFactory = null
+        \Magento\Catalog\Model\ResourceModel\Eav\AttributeFactory $attributeFactory = null,
+        \Magento\Catalog\Model\ProductRepository $mediaGalleryProcessor,
+//        \Magento\Catalog\Api\Data\ProductInterface $productInterface,
```
Comment (engcom-Hotel) [inline app/code/Magento/ConfigurableProduct/Model/LinkManagement.php]: I think we can remove it from here as it is of no use

File: app/code/Magento/ConfigurableProduct/Model/LinkManagement.php [hunk 3]
```diff
@@ -55,24 +65,47 @@ class LinkManagement implements \Magento\ConfigurableProduct\Api\LinkManagementI
...
+        \Magento\Catalog\Api\Data\ProductAttributeMediaGalleryEntryInterfaceFactory $myModelFactory
     ) {
         $this->productRepository = $productRepository;
         $this->productFactory = $productFactory;
         $this->configurableType = $configurableType;
         $this->dataObjectHelper = $dataObjectHelper;
         $this->attributeFactory = $attributeFactory ?: \Magento\Framework\App\ObjectManager::getInstance()
             ->get(\Magento\Catalog\Model\ResourceModel\Eav\AttributeFactory::class);
+        $this->mediaGallery = $mediaGalleryProcessor;
+//        $this->productInterface=$productInterface;
```
Comment (engcom-Hotel) [inline app/code/Magento/ConfigurableProduct/Model/LinkManagement.php]: I think we can remove it from here as it is of no use

**Review by engcom-Hotel — CHANGES_REQUESTED** [review]
Hello @JettyRohanAditya,

Thanks for the contribution!

Please review the comments below and add some automated tests to cover the PR changes. Also please resolve the build failures.

Thanks

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

File: app/code/Magento/ConfigurableProduct/Model/LinkManagement.php [hunk 4]
```diff
@@ -55,24 +64,45 @@ class LinkManagement implements \Magento\ConfigurableProduct\Api\LinkManagementI
...
      */
     public function __construct(
         \Magento\Catalog\Api\ProductRepositoryInterface $productRepository,
         \Magento\Catalog\Api\Data\ProductInterfaceFactory $productFactory,
         \Magento\ConfigurableProduct\Model\ResourceModel\Product\Type\Configurable $configurableType,
         \Magento\Framework\Api\DataObjectHelper $dataObjectHelper,
-        \Magento\Catalog\Model\ResourceModel\Eav\AttributeFactory $attributeFactory = null
+        \Magento\Catalog\Model\ResourceModel\Eav\AttributeFactory $attributeFactory = null,
+        \Magento\Catalog\Model\ProductRepository $mediaGalleryProcessor,
+        \Magento\Catalog\Api\Data\ProductAttributeMediaGalleryEntryInterfaceFactory $myModelFactory
```
Comment (engcom-Hotel) [inline app/code/Magento/ConfigurableProduct/Model/LinkManagement.php]: Adding Params in this way may be backward incompatible, set them null as default and use objectmanager to initialize those later. Please refer to the `$this->attributeFactory` initialization for reference.

**Review by engcom-Hotel — CHANGES_REQUESTED** [review]

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run static tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run Functional Tests CE

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE, Integration Tests, WebAPI Tests

**Review by engcom-Hotel — APPROVED** [review]

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento give me test instance

**Comment by engcom-Delta** [comment]
### :heavy_check_mark: QA Passed
 
**Preconditions:**
- Install fresh Magento **`2.4-develop`** and **`PHP 8.1`**
- Postman should be configured with magento2
 
**Manual testing scenario:**

	1	Create a new configurable product
	2	Add configurations with unique product images
	3	Make a request in postman using the /configurable-products/{sku}/children


**Before:** :heavy_multiplication_x: 
before 37862
 
 
**After:** :heavy_check_mark:  
 after 37862
 

Builds are failed. Hence, moving this PR to Extended Testing.

**Comment by engcom-Echo** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Echo** [comment]
Functional Tests B2B failure are different on last two run on same commit.

Run1
Screenshot 2023-10-04 at 10 13 12 AM

Run2
Screenshot 2023-10-04 at 10 13 19 AM

Hence moving it to `merge in progress`

**Comment by engcom-Echo** [comment]
Functional Tests B2B failure are different on last two run on same commit.

Run1
Screenshot 2023-10-04 at 10 13 12 AM

Run2
Screenshot 2023-10-04 at 10 13 19 AM

Hence moving it to `merge in progress`

**Pull request merged**